lib/trace.c

```diff
@@ -148,18 +148,24 @@ M0_INTERNAL void m0_trace_fini(void)
 	m0_arch_trace_fini();
 }
 
-
-/*
- * XXX x86_64 version.
- */
+/* TODO in case user mode support required for AARCH64 different mechanism is needed */
 static inline uint64_t m0_rdtsc(void)
```

Comment:
Sorry, I don't understand - isn't this function compiled and used in user-mode?

Reply:
This is suggested by vemuri to add the comments

Reply:
it is kernel mode  function

Reply:
As far as I can see - it is used in both: kernel *ant* user-space mode.

Reply:
I mean if kernel module is not there, this code won't work for AARCH.

Reply:
No in kernel mode only

Reply:
From the comment, it's not quite clear why. Please elaborate your comment like this: "pccnt access from user-space in AARCH64 platform is enabled by enable_user_space_access_pccnt() call from the kernel module, so this function will work only if the kernel module is present on this platform. TODO: investigate how to resolve it for the user-space mode in case there is no kernel module present."

Reply:
This is privileged instruction for enabling the cycle counter register. which is done in kernel mode but user can read the performance counter in user mode